### PR TITLE
[github #4555][devops #1761663] [Bug fixes] Azure CLI authentication does not show subscriptions for all tenants

### PR DESCRIPTION
**Pull request by wangmingliang-ms**

### CODE REVIEW

**Review by Zenix27 — APPROVED** [review]

File: Utils/azuretools-core/src/com/microsoft/azuretools/utils/CommandUtils.java [hunk 1]
```diff
@@ -56,24 +64,51 @@
...
+        final ByteArrayOutputStream err = new ByteArrayOutputStream();
+        final PumpStreamHandler streamHandler = new PumpStreamHandler(out, err);
         final DefaultExecutor executor = new DefaultExecutor();
         executor.setWorkingDirectory(directory);
         executor.setStreamHandler(streamHandler);
         executor.setExitValues(null);
         try {
             executor.execute(commandLine);
-            return outputStream.toString();
+            System.err.println(err.toString());
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azuretools/utils/CommandUtils.java:104]: should log other than print error message.

File: Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/ServicePrincipalAzureManager.java [hunk 2]
```diff
@@ -198,55 +78,65 @@ public KeyVaultClient getKeyVaultClient(String tid) {
...
 
-    private void initATCIfNeeded() throws IOException {
-        if (atc == null) {
-            atc = ApplicationTokenCredentials.fromFile(credFile);
+    private void init() {
+        if (this.credentials == null && credFile != null) {
+            try {
+                this.credentials = ApplicationTokenCredentials.fromFile(credFile);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/ServicePrincipalAzureManager.java:127]: avoid to convert checked exception to unchecked ones, use AzureExecutionException

File: Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/AzureCliAzureManager.java [hunk 3]
```diff
@@ -22,193 +22,112 @@
...
+     *       identity/implementation/IdentityClient.java#L366
+     */
+    private Pair<String, OffsetDateTime> getAccessTokenViaCli(String tid) {
+        //
+        final String command = String.format("az account get-access-token --output json -t %s", tid);
+        final String jsonToken;
+        try {
+            jsonToken = CommandUtils.exec(command);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/AzureCliAzureManager.java:100]: throw IOException

File: Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/AzureCliAzureManager.java [hunk 4]
```diff
@@ -22,193 +22,112 @@
...
-    @Override
-    public KeyVaultClient getKeyVaultClient(String tid) {
-        if (!isSignedIn()) {
-            return null;
+    /**
+     * refer https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/identity/azure-identity/src/main/java/com/azure/
+     *       identity/implementation/IdentityClient.java#L366
+     */
+    private Pair<String, OffsetDateTime> getAccessTokenViaCli(String tid) {
+        //
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/AzureCliAzureManager.java:94]: remove empty //

Comment (andxu) [reply Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/AzureCliAzureManager.java:94]: check id for valid guid

File: Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/AzureCliAzureManager.java [hunk 5]
```diff
@@ -240,11 +163,35 @@ public AuthMethodDetails signIn() throws AzureExecutionException {
         }
     }
 
-    private Azure.Authenticated auth() {
-        return Azure.configure()
-                .withInterceptor(new TelemetryInterceptor())
-                .withUserAgent(CommonSettings.USER_AGENT)
-                .authenticate(azureCliCredentials);
+    private static <V> Map<String, V> convertJsonToMap(@NotNull String jsonString) {
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/AzureCliAzureManager.java:166]: use JsonUtils?

File: Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/AzureManagerBase.java [hunk 6]
```diff
@@ -96,4 +115,181 @@ public String getTenantIdBySubscription(String subscriptionId) throws IOExceptio
...
+    }
+
+    @Override
+    public Settings getSettings() {
+        return settings;
+    }
+
+    @Override
+    public void drop() throws IOException {
+        System.out.println("ServicePrincipalAzureManager.drop()");
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/AzureManagerBase.java:232]: avoid System.out.println, use log.info

File: Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/AzureCliAzureManager.java [hunk 7]
```diff
@@ -22,193 +22,112 @@
...
 
-    @Override
-    public KeyVaultClient getKeyVaultClient(String tid) {
-        if (!isSignedIn()) {
-            return null;
+    /**
+     * refer https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/identity/azure-identity/src/main/java/com/azure/
+     *       identity/implementation/IdentityClient.java#L366
+     */
+    private Pair<String, OffsetDateTime> getAccessTokenViaCli(String tid) {
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/AzureCliAzureManager.java:93]: move private methods last

File: Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/AzureCliAzureManager.java (hunk 3 above)
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/AzureCliAzureManager.java:100]: void convert IOException  to RuntimeException, add throws if neccessary

File: Utils/azuretools-core/src/com/microsoft/azuretools/utils/CommandUtils.java (hunk 1 above)
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azuretools/utils/CommandUtils.java:104]: avoid to call System.err.println, use LOGGER.error

File: Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/AzureCliAzureManager.java (hunk 4 above)
Comment (wangmingliang-ms) [reply Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/AzureCliAzureManager.java]: do we have to check if the `tid` is valid GUID? is `tid` certainly a valid GUID?

**Comment by andxu** [comment]
ON-hold this PR to next sprint

File: Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/AzureCliAzureManager.java [hunk 8]
```diff
@@ -58,136 +57,53 @@
     private static final String FAILED_TO_AUTH_WITH_AZURE_CLI = "Failed to auth with Azure CLI";
     private static final String UNABLE_TO_GET_AZURE_CLI_CREDENTIALS = "Unable to get Azure CLI credentials, " +
             "please ensure you have installed Azure CLI and signed in.";
+    public static final String CLI_TOKEN_FORMAT_ACCESSOR = "az account get-access-token --output json -t %s";
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azuretools/sdkmanage/AzureCliAzureManager.java:60]: public -> private

**Review by andxu — APPROVED** [review]

**Pull request merged**